### PR TITLE
[Cache] Advanced Memcached Adapter

### PR DESCRIPTION
**Pull request by robfrawley**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | **yes**
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | **yes**
| Related Issues | symfony/symfony-docs#7256
| License       | MIT
| Doc PR        | symfony/symfony-docs#7265
| Related PRs | #20858, ~~#20752~~

Expansion of simple adapter from #20858. Replacement for #20752.

### CODE REVIEW

**Comment by robfrawley** [comment]
@nicolas-grekas @stof This PR vastly cuts down on the server/option logic from prior proposals. It builds on #20858

**Comment by robfrawley** [comment]
This has been updated with any changes made against #20858

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        }
+
+        return array_filter((array) $ids, function ($id) use ($namespace) {
+            return 0 === strpos($id, $namespace);
+        });
+    }
+
+    private function setOption($opt, $val)
+    {
+        $toRestore = error_reporting(~E_ALL);
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: why changing the error reporting ?

Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: To return a false instead of an E_WARNING on invalid option passed.

Comment (stof) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: Well, IMO, the warning would be much easier to debug, as it explains what is wrong. So I'm -1 on this.

Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: See https://github.com/symfony/symfony/pull/20863/files#diff-ed071d63655500d46a7c7866cdb74cddR193 for test checking a false return on error instead of an E_WARNING. Is this something you don't think we should do?

Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: OR, another option; continue catching the E_WARNNG but transform into an exception, which IMHO, may be the most developer friendly option?

Comment (stof) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: Well, we could just let the error handler do it, without doing any special handling for E_WARNING happening inside ``setOption``

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        return $adapter;
+    }
+
+    /**
+     * @param string[] $dsns
+     * @param mixed[]  $opts
+     *
+     * @return bool
+     */
+    public function setup(array $dsns = array(), array $opts = array())
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: I don't think we should expose such public API

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        $adapter->setup($dsn ? array($dsn) : array(), $opts);
+
+        return $adapter;
+    }
+
+    /**
+     * @param string[] $dsns
+     * @param mixed[]  $opts
+     *
+     * @return bool
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: Returning a boolean does not make sense here:

- handling errors by boolean return values is not user-friendly (no message explaining the issue)
- it is ignored by ``create()`` anyway, meaning you don't have error handling at all

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+        if (isset($srv['query']) && 1 === preg_match('{weight=([^&]{1,})}', $srv['query'], $weight)) {
+            unset($srv['query']);
+            $srv['weight'] = (int) $weight[1];
+        }
+
+        return array_values($srv);
+    }
+
+    protected function isServerInClientPool($host, $port)
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: why protected ?

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    {
+        return (bool) array_filter($this->client->getServerList(), function ($srv) use ($host, $port) {
+            return $host === array_shift($srv) && $port === array_shift($srv);
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: you should avoid shifting the array. Mutating an array for reading it is a bad idea: it forces PHP to copy it (not that bad here as it has only a few items, but very bad in general

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+            unset($srv['query']);
+            $srv['weight'] = (int) $weight[1];
+        }
+
+        return array_values($srv);
+    }
+
+    protected function isServerInClientPool($host, $port)
+    {
+        return (bool) array_filter($this->client->getServerList(), function ($srv) use ($host, $port) {
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: using array_filter is a bad idea: it forces to compare all servers even if the first one is a match, meaning that you force to always reach the worse case (when there is no match, we are indeed forced to compare them all)

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        return array(
+            'OPT_LIBKETAMA_COMPATIBLE' => true,
+            'OPT_COMPRESSION' => true,
+        );
+    }
+
+    public static function setupBeforeClass()
+    {
+        if (!MemcachedAdapter::isSupported()) {
+            self::markTestSkipped('Extension memcached >2.1.0 required.');
```
Comment (stof) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: the min version is 2.2.0 in the code, not 2.1.0

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+            static::defaultClientOptions()
+        );
+        static::$client = static::getClientFromAdapter($adapter);
+
+        parent::setupBeforeClass();
+    }
+
+    public static function tearDownAfterClass()
+    {
+        self::$client->flush();
```
Comment (stof) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: shouldn't it be flushed between tests instead of at the end of the testcase, to ensure test isolation ?

I would even go further and avoid making the client static, ensuring we don't reuse it between tests at all.

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        self::$client = null;
+    }
+
+    public function createCachePool($defaultLifetime = 0)
+    {
+        return new MemcachedAdapter(self::$client, str_replace('\\', '.', __CLASS__), $defaultLifetime);
+    }
+
+    public function testCreateAdaptor()
```
Comment (stof) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: shouldn't it be ``adapter`` ?

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php (hunk 1 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: Removed error suppression (https://github.com/symfony/symfony/pull/20863/commits/2f3aeedc58fea85f488641181e7c957525b2a99c#diff-e01499fab5ad7f5e36517d157bc13e2fR155) and amended test to check for E_WARNING (https://github.com/symfony/symfony/pull/20863/commits/2f3aeedc58fea85f488641181e7c957525b2a99c#diff-ed071d63655500d46a7c7866cdb74cddR177).

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php (hunk 7 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: `>2.1.0` and `>=2.2.0` are the same thing, but I follow the logic of clarifying this. :+1:

Comment (stof) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: not the same. ``2.2.0RC1`` is ``> 2.1.0`` but is not ``>= 2.2.0`` (the same would be true for ``2.1.1`` if there was such a release)

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php (hunk 6 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: Changed to foreach: https://github.com/symfony/symfony/pull/20863/commits/3947b38f7352a310e83e5794fc3bbff61e94e2f0#diff-e01499fab5ad7f5e36517d157bc13e2fR191

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php (hunk 5 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: Gone

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php (hunk 7 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: Clarified: https://github.com/symfony/symfony/pull/20863/commits/3947b38f7352a310e83e5794fc3bbff61e94e2f0#diff-ed071d63655500d46a7c7866cdb74cddR49

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php (hunk 3 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: Returns `$this` now: https://github.com/symfony/symfony/pull/20863/commits/3947b38f7352a310e83e5794fc3bbff61e94e2f0#diff-e01499fab5ad7f5e36517d157bc13e2fR72

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php (hunk 2 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: This was a middle-ground approach to allow a simple string DSN for `create` but allow for an array of DSNs using a `calls` during service construction to this method.

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php (hunk 9 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: Yup ;-) But that test is actually gone now.

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php (hunk 8 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: This is how the Redis adapter handles it, but yeah; makes sense. Changed.

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php (hunk 2 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: As well, without this method, all the logic from [setup()](https://github.com/symfony/symfony/pull/20863/commits/f780643c781eee70e1516eeee40345049599299f#diff-e01499fab5ad7f5e36517d157bc13e2fR72), [setOption()](https://github.com/symfony/symfony/pull/20863/commits/f780643c781eee70e1516eeee40345049599299f#diff-e01499fab5ad7f5e36517d157bc13e2fR152), [resolveOption()](https://github.com/symfony/symfony/pull/20863/commits/f780643c781eee70e1516eeee40345049599299f#diff-e01499fab5ad7f5e36517d157bc13e2fR160), [addServer()](https://github.com/symfony/symfony/pull/20863/commits/f780643c781eee70e1516eeee40345049599299f#diff-e01499fab5ad7f5e36517d157bc13e2fR165), [parseServerDSN()](https://github.com/symfony/symfony/pull/20863/commits/f780643c781eee70e1516eeee40345049599299f#diff-e01499fab5ad7f5e36517d157bc13e2fR174), and [isServerInClientPool()](https://github.com/symfony/symfony/pull/20863/commits/f780643c781eee70e1516eeee40345049599299f#diff-e01499fab5ad7f5e36517d157bc13e2fR191) would need to be shoved into [create()](https://github.com/symfony/symfony/pull/20863/commits/f780643c781eee70e1516eeee40345049599299f#diff-e01499fab5ad7f5e36517d157bc13e2fR58).

**Comment by robfrawley** [comment]
Latest commits bring in line with latest changes to #20858

**Comment by robfrawley** [comment]
Removed all instanced of `array_shift` (there were three)...

**Comment by robfrawley** [comment]
Squashed commits. All changes from #20858 implemented in this PR.

File: src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php (hunk 8 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Tests/Adapter/MemcachedAdapterTest.php]: The base test class already calls clear on the pool, that's why it's not needed, isn't it?

**Comment by nicolas-grekas** [comment]
Rebase needed

**Comment by robfrawley** [comment]
@nicolas-grekas done!

**Comment by robfrawley** [comment]
Also, amended this commit to be in-line with @fabpot style fixes from https://github.com/symfony/symfony/commit/51d13cc73ab8d6b6330789e8740b19d7b00a1796 on my first Memcached PR #20858

**Comment by robfrawley** [comment]
Test failures unrelated to this PR: https://travis-ci.org/symfony/symfony/jobs/183666262#L2697

**Comment by robfrawley** [comment]
I don't know how I feel about the [second commit](https://github.com/symfony/symfony/pull/20863/commits/6819e1c60d850bd497ffbe95cb806ae6a7c7ffb9). I don't think it's really necessary, but wanted feedback on whether it could be useful or not.

**Comment by nicolas-grekas** [comment]
So, I promised you many comments after #20858, here they are :)

There should be only one single `createClient` method. This method should return a Memcached instance. I will make the methods much more powerful (could be used in more use cases), and make the instance stronger (no mutability of the connection).
`createClient` should take two arguments: a DSN string, and an array of options.
the DSN should handle everything that is the responsility of the sysops.
the options array should handle everything that is the responsility of the appdev.
for this reason, the current values in `$clientDefaultServer` make no sense to me, because they are not *behavioral* defaults, relates to sysops things, which can't have a "good" default by definition.

The createConnection method of the RedisAdapter has been written with this design in mind. In invite you to read it with these eyes.

**Comment by robfrawley** [comment]
@nicolas-grekas At a very basic level (mind you, this is UGLY code, but I just want to make sure we're on the same page), you want something like this:

```php
<?php

require_once "/home/rmf/repositories/symfony/symfony/vendor/autoload.php";

$servers = array(
  'memcached://localhost',
  'memcached://localhost:11211',
  'memcached://127.0.0.1:11211',
  'memcached://localhost:11211?persistent_id=foo',
  'memcached://localhost:11211/?weight=50',
  'memcached://localhost:11211?persistent_id=foo&weight=50',
  'memcached://localhost:11211/?weight=50&persistent_id=foo',
  'memcached:///var/run/memcached.sock',
  'memcached:///var/run/memcached.sock?persistent_id=foo',
  'memcached:///var/run/memcached.sock/?weight=50',
  'memcached:///var/run/memcached.sock?persistent_id=foo&weight=50',
  'memcached:///var/run/memcached.sock/?weight=50&persistent_id=foo',
);

$defaults = [
    'host' => 'localhost',
    'port' => 11211,
    'weight' => 100,
    'persistent_id' => null,
];

$configs = [];

foreach ($servers as $i => $s) {
    $configs[$i]['dsn'] = $s;

    if (false !== $url = parse_url($s)) {
        parse_str(isset($p['query']) ? $p['query'] : '', $q);

        $configs[$i]['type'] = 'TCP';
        $configs[$i] += $url;
        unset($configs[$i]['path']);
        unset($configs[$i]['scheme']);

        parse_str(isset($p['query']) ? $p['query'] : '', $q);
        if ($q) {
            $configs[$i] += $q;
        }

        $configs[$i] += $defaults;

        continue;
    }

    if (1 !== preg_match('{memcached:\/\/(?<host>\/[^?]+)(?:\?)?(?<query>.+)?}', $s, $m)) {
        throw new \Exception('BAD: '.$s);
    }

    $configs[$i]['type'] = 'SOCKET';
    $configs[$i]['host'] = $m['host'];

    parse_str(isset($m['query']) ? $m['query'] : '', $q);
    if ($q) {
        $configs[$i] += $q;
    }

    $configs[$i] += $defaults;
    unset($configs[$i]['port']);
}

$configs = array_map(function ($s) {
    uksort($s, function ($a, $b) {
        return $a > $b;
    });
    array_walk($s, function (&$v, $i) {
        if ($i === 'weight') {
            $v = (int) $v;
        }
        if ($i === 'host' && strpos(strrev($v), '/') === 0) {
            $v = substr($v, 0, -1);
        }
    });
    return $s;
}, $configs);

echo \Symfony\Component\Yaml\Yaml::dump($configs);
```

Which give us:

```yml
-
    dsn: 'memcached://localhost'
    host: localhost
    persistent_id: null
    port: 11211
    type: TCP
    weight: 100
-
    dsn: 'memcached://localhost:11211'
    host: localhost
    persistent_id: null
    port: 11211
    type: TCP
    weight: 100
-
    dsn: 'memcached://127.0.0.1:11211'
    host: 127.0.0.1
    persistent_id: null
    port: 11211
    type: TCP
    weight: 100
-
    dsn: 'memcached://localhost:11211?persistent_id=foo'
    host: localhost
    persistent_id: null
    port: 11211
    query: persistent_id=foo
    type: TCP
    weight: 100
-
    dsn: 'memcached://localhost:11211/?weight=50'
    host: localhost
    persistent_id: null
    port: 11211
    query: weight=50
    type: TCP
    weight: 100
-
    dsn: 'memcached://localhost:11211?persistent_id=foo&weight=50'
    host: localhost
    persistent_id: null
    port: 11211
    query: 'persistent_id=foo&weight=50'
    type: TCP
    weight: 100
-
    dsn: 'memcached://localhost:11211/?weight=50&persistent_id=foo'
    host: localhost
    persistent_id: null
    port: 11211
    query: 'weight=50&persistent_id=foo'
    type: TCP
    weight: 100
-
    dsn: 'memcached:///var/run/memcached.sock'
    host: /var/run/memcached.sock
    persistent_id: null
    type: SOCKET
    weight: 100
-
    dsn: 'memcached:///var/run/memcached.sock?persistent_id=foo'
    host: /var/run/memcached.sock
    persistent_id: foo
    type: SOCKET
    weight: 100
-
    dsn: 'memcached:///var/run/memcached.sock/?weight=50'
    host: /var/run/memcached.sock
    persistent_id: null
    type: SOCKET
    weight: 50
-
    dsn: 'memcached:///var/run/memcached.sock?persistent_id=foo&weight=50'
    host: /var/run/memcached.sock
    persistent_id: foo
    type: SOCKET
    weight: 50
-
    dsn: 'memcached:///var/run/memcached.sock/?weight=50&persistent_id=foo'
    host: /var/run/memcached.sock
    persistent_id: foo
    type: SOCKET
    weight: 50
```

**Comment by robfrawley** [comment]
@nicolas-grekas So... I took this is a bit of a different direction, as the idea of huge factory methods seems counter to providing a clear, clean, and developer friendly codebase (such as the many `createConnection` (or equivalent) methods that exist, like the `RedisAdapter` one).

Take a look at my idea for abstracting this logic outside of the actual adapter. If such seems like an approach you like, I'd be more than happy to submit PRs to do the same for other adapters as well to clean them up a bit.

**Comment by robfrawley** [comment]
Test failures unrelated to this PR: https://travis-ci.org/symfony/symfony/jobs/183705594#L2729

File: src/Symfony/Component/Cache/Adapter/Client/ClientInterface.php [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ */
+interface ClientInterface
+{
+    public static function isSupported();
+
+    public static function getRequirements();
+
+    public static function create($servers = array(), array $options = array());
+
+    public function getClient();
```
Comment (stof) [inline src/Symfony/Component/Cache/Adapter/Client/ClientInterface.php]: this makes no sense, as you are unable to document the type of this method

**Comment by stof** [comment]
IMO, you should not have this Client concept. It is just unnecessary.

If people want full power over the configuration of their Memcached instance, they already have it, but doing the configuration themselves and passing the instance to the adapter constructor. This is what dependency injection is about

**Comment by robfrawley** [comment]
@stof 

The issue, though, is that there *will be __some__ configuration logic somewhere* to allow for a factory method like [`Redis::createConnection()`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Adapter/RedisAdapter.php#L68) for the `MemcachedAdapter`, which is the requested concept from @nicolas-grekas (a concept I agree with, as well). What I don't agree with is the current idea of shoving all the required logic into that single method. Instead of having a [`63` line behemoth (for `RedisAdapter`)](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Adapter/RedisAdapter.php#L68), that is, frankly, exceptionally hard to follow, it would make more sense to move this logic into a factory class, so we **can break out the logic into easy to follow, maintain, and DX-friendly methods in a client setup class**.

Now, **specifically with regard to `\Memcached`**, providing some factory for client configuration **is a _requirement_ for the user**, if one isn't otherwise provided. This is due to a bunch of odd and unexpected behavior exhibited by the extension, as well as a generally DX-unfriendly interface.

For example, take the following DI service configuration a user might use with this adapter:

```yaml
app.cache.memcached.client:
    class: \Memcached
    arguments:
        - my_persistent_connection_id
    calls:
        - [setOptions, [[{"!php/const:\Memcached::OPT_SERIALIZER": "!php/const:\Memcached::SERIALIZER_IGBINARY", "!php/const:\Memcached::OPT_LIBKETAMA_COMPATIBLE": true}]]]
        - [setServers, [[["localhost", 11211, 50], ["localhos", 11212, 50]]]]
app.cache:
    class: \Symfony\Component\Cache\Adapter\MemcachedAdapter
    arguments:
        - "@app.cache.memcached.client"
        - "my_namespace"
        - 3600
```

Now, ignoring the fact that **I don't know if that will even parse properly** (due to the associative array of constants), I think everyone would agree *it is not user-friendly or particularly easy to follow or write*. But the issues don't end there. Due to the way the extension handles the `setServers` method, you will end up with 200 servers registered after 100 web requests, as the servers are persisted in different environments (for example, FPM); the above will re-add the servers every time the container instantiates this service.

In my mind, we want to provide the user with easy to use helpers to allow for simple usage of the code and protect against commonly-misused (or mis-understood/unknown) quirks they would otherwise encounter.

Given the proposed code, the service definition would be:

```yaml
app.cache.memcached.client:
    class: Symfony\Component\Cache\Adapter\Client\MemcachedClient
    arguments:
        - ["memcached://localhost:11211?weight=50", "memcached://localhost:11212?weight=50"]
        - {serializer: "igbinary", libketama_compatible: true}
```

Given the above, the user is protected from server duplications and provided a **much** cleaner wrapper to creating a \Memcached instance.

File: src/Symfony/Component/Cache/Adapter/Client/ClientInterface.php (hunk 10 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/Client/ClientInterface.php]: Absolutely true. If you ignore that you don't like the idea of providing a wrapper for configuring `\Memcached`, would you have suggestions as to an interface that *would* make sense?

**Comment by nicolas-grekas** [comment]
I missed explaining why we need this conceptual separation between a DSN for sysops and an array of options for appdevs, and why we need a DSN *string* in the first place.
The reason is that a DSN string fits in a single environment variable.
That's why e.g. having an array of DSN for several severs won't help: a sysops need to be able to configure the Memcached server in a single environment variable.
Everything I expressed as requirement derives from this idea.

**Comment by robfrawley** [comment]
The proposed code allows a string or array of DSNs.

**Comment by nicolas-grekas** [comment]
Let's go for string or array of DSN.
But really, we only need *one* public method. We don't need nor want a single more.
You have to consider that each new public/protected/interface creates a maintainability burden on the community to keep the BC promise.
In this case, all the other added methods are just helpers. There's no way we're going to want to maintain these.

**Comment by robfrawley** [comment]
Removed the other interface methods, except for [`isSupported()`](https://github.com/symfony/symfony/pull/20863/files#diff-ce74fa255a89e743bcb9a61ce44dcef3R19) which IMO seems like a much more applicable place to have this check (in [the *Client class](https://github.com/symfony/symfony/pull/20863/files#diff-efa493248313205dd52a55f253fa4fe0R41) than within the adapter, and is likely applicable to all adapter clients, right? If not, give me a heads up and I'll get rid of it.

**Comment by nicolas-grekas** [comment]
Yes, please remove AbstractClient and the interface. Please also make sure that the createClient (or createConnection) returns a *Memcached* instance.
Wrapping it in a "new static" *reduces* the usefulness of the method.

**Comment by robfrawley** [comment]
Makes sense (re: \Memcached return). Interface/abstract removed as well!

File: src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+ */
+
+namespace Symfony\Component\Cache\Adapter\Client;
+
+use Symfony\Component\Cache\Exception\InvalidArgumentException;
+
+/**
+ * @author Rob Frawley 2nd <rmf@src.run>
+ */
+class MemcachedClient
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php]: This class should be removed: creating a Memcached layer is not the responsibility of the Cache component. But by having this class, people *will* use it as such. An alternative could be to mark it as internal, but I'm not sure this makes more sense.

File: src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+{
+    private static $serverDefaults = array(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php]: I think we don't need server defaults in the cache component (we may have in FrameworkBundle later on, but that's another topic.)

File: src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+        'host' => 'localhost',
+        'port' => 11211,
+        'weight' => 100,
+    );
+
+    private static $optionDefaults = array(
+        'compression' => true,
+        'libketama_compatible' => true,
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php]: What do you think about enabling binary mode by default?

File: src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+            return false;
+        }
+
+        return $this->resolveServerReturn(array(
+            'host' => 0 === strpos(strrev($matches['host']), '/') ? substr($matches['host'], 0, -1) : $matches['host'],
+            'query' => isset($matches['query']) ? $matches['query'] : null,
+        ));
+    }
+
+    private function resolveServerReturn($server)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php]: What about username+ password for sasl?

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php [hunk 15]
```diff
@@ -75,4 +83,13 @@ protected function doClear($namespace)
...
         return $this->client->flush();
     }
+
+    public function __destruct()
+    {
+        if (!$this->client->isPersistent() && method_exists($this->client, 'flushBuffers')) {
+            $this->client->flushBuffers();
+        }
+
+        parent::__destruct();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: Shouldn't the parent be called before?

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php [hunk 16]
```diff
@@ -75,4 +83,13 @@ protected function doClear($namespace)
     {
         return $this->client->flush();
     }
+
+    public function __destruct()
+    {
+        if (!$this->client->isPersistent() && method_exists($this->client, 'flushBuffers')) {
+            $this->client->flushBuffers();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: This is not done automatically when the connection is destructed?

**Review by nicolas-grekas — COMMENTED** [review]
We only need one single new public method in the whole PR :)

File: src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php (hunk 11 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php]: Marked `@internal` for now. I **really, really, really** want to avoid shoving all this logic into a single `createConnection` method within the Adapter itself...

File: src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php (hunk 12 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/Client/MemcachedClient.php]: These are the distribution defaults for an installed memcached server with **no user configuration applied**. Additionally, this doesn't actually **add a default server** if no DSN is passed. At a minimum, the user still needs to pass `memcached:` as the DSN. An empty array or empty string will **not** add a server.

File: src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php (hunk 16 above)
Comment (robfrawley) [inline src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php]: TBH, this method isn't even documented yet, found it when looking through the extension source to determine available options for the documentation PR. It looks like this is supposed to be called if the new `buffer_writes` option is enabled.

**Comment by robfrawley** [comment]
Also, made the `MemcachedClient` `final`, as well as `@internal`

**Comment by robfrawley** [comment]
Documentation PR submitted (assuming this is merged as-is) as symfony/symfony-docs#7265: [view it here](http://pr-7265-vfyqlyy-6qmocelev2lwe.eu.platform.sh/components/cache/adapters/memcached_adapter.html).

**Comment by robfrawley** [comment]
Why is platform.sh [ignoring `markTestSkipped`](https://github.com/symfony/symfony/pull/20863/files#diff-4e51ce1d6022f4eaf2e3548cf4d6d799R21) [causing this test failure](https://travis-ci.org/symfony/symfony/jobs/184824689#L3482)?

**Comment by nicolas-grekas** [comment]
Continued in #21108
Thank you @robfrawley!
